### DISCUSSION TITLE
Define multiple IntlMessages types for use in monorepo

### DISCUSSION THREAD

**Discussion by jonathanwilke** [amannn/next-intl#1224] [Q&A]
Hey there,

I am using next-intl in a monorepo and I am trying to get next-intl running in a lib there that defines my mail templates with react-email. I have two global.d.ts files that define IntlMessages, one in my main project (a Next.js app) and one in the library for the mail templates. The main app imports the mail templates from the library to send the mails. In the editor everything works fine with the types, but if I try to build the app, I get the error that the messages I pass to the createTranslator method are not valid in the mail template library. I suppose this is due to the fact the it takes the IntlMessages type from my main app, which of course does not match the messages of the mail template library.

Is there any way how I can define override the types for a specific instance of the createTranslator method or any other way to fix this?

**Comment by amannn** [comment]
Hmm, sounds like an issue with how types are used in the respective packages? Can you maybe skip the type check of the email package when building the main app? Sounds like something you might want to do anyway for perf reasons, as long as the email package takes care of checking its types internally?

Maybe something like [`skipLibCheck`](https://www.typescriptlang.org/tsconfig/#skipLibCheck) does the trick?

**Reply by jonathanwilke** [reply]
I solved it (temporarilly) now by just putting everything into one language file. This would of course not be a problem if we had scoped translation files.

**Reply by amannn** [reply]
This sounds mostly like a TypeScript tooling issue, no? I'd say ideally you should have different messages for your email and for your main app package. Each package should be able to define messages and make sure types don't leak into other packages.

**Reply by jonathanwilke** [reply]
The problem is that the main package imports the other one and at build time this will crash due to type errors

**Reply by flipflopsandrice** [reply]
I have a similar issue, I'm using TurboRepo. 

Let me explain:

**package/somename**
- `Translations` package specific type in a `translations.d.ts` file, exported from the package
- Has components with `useTranslations`: :heavy_check_mark:  _type checking works on build_

**package/otherone**
- `Translations` package specific type in a `translations.d.ts` file, exported from the package
- Has components with `useTranslations`: :heavy_check_mark: _type checking works on build_

**apps/main:**
- Imports `Translations` types from `package/somename` and `package/otherone` in a `global.d.ts` and extend the `IntlMessages` interface with a union of the types.
- Has components with `useTranslations`:  :heavy_check_mark: _type checking works on build_
- Import components (`package/somename`) with `useTranslations`:  **:warning: _type checking breaks on build_**

---

**apps/main/global.d.ts**
```
import type { SomenameTranslations } from '@repo/somename/types/translations';
import type { OtheroneTranslations } from '@repo/otherone/types/translations';

declare global {
  interface IntlMessages extends Translations {}
}

export {};
```

**tsc error**
```
main:types: ../../packages/somename/src/components/MyComponent.tsx:22:24 - error TS2345: Argument of type 'string' is not assignable to parameter of type 'never'.
main:types: 
main:types: 22       myPropertyTitle: t('somename.myproperty.title'),
```

**Reply by gethari** [reply]
@flipflopsandrice Did you solve this ? I am facing the exact same issue
